### PR TITLE
 Added Hutch++ trace estimator and tests for trace estimators

### PR DESCRIPTION
**Pull request by oliversssf2**
1. Implemented Hutch++
2. Wrote sanity-check tests for trace estimators

TODOs:
1. Implement empirical variance for Hutchinson and Hutch++
2. Implement statistical tests for trace estimators

### CODE REVIEW

File: rlaopt/spectral_estimators/trace.py [hunk 1]
```diff
@@ -9,16 +9,53 @@
...
+    """
+    Omega = get_sketch(sketch, "left", k, A.shape[0], torch.float32, A.device)
     Omega_A = Omega._apply_left(A)
     Omega_A_Omega_T = Omega._apply_right_trans(Omega_A)
     d = Omega_A_Omega_T.diag()
-    trace = torch.sum(d)
-    var = 1 / (k - 1) * torch.sum(k * d - trace)
-    return trace, var
+    trace = torch.sum(d) / k
+    # var = 1 / (k - 1) * torch.sum(k * d - trace)
```
Comment (pratikrathore8) [inline rlaopt/spectral_estimators/trace.py:28]: I'm ok with not returning the variance -- let's just remove the line that has been commented out

File: rlaopt/spectral_estimators/trace.py [hunk 2]
```diff
@@ -9,16 +9,53 @@
...
+        float: Estimated trace of A.
+    """
+    Omega = get_sketch(sketch, "left", k, A.shape[0], torch.float32, A.device)
     Omega_A = Omega._apply_left(A)
     Omega_A_Omega_T = Omega._apply_right_trans(Omega_A)
     d = Omega_A_Omega_T.diag()
-    trace = torch.sum(d)
-    var = 1 / (k - 1) * torch.sum(k * d - trace)
-    return trace, var
+    trace = torch.sum(d) / k
```
Comment (pratikrathore8) [inline rlaopt/spectral_estimators/trace.py:27]: I think we're better off using the built-in `torch.mean`

File: rlaopt/spectral_estimators/trace.py [hunk 3]
```diff
@@ -9,16 +9,53 @@
...
+    # var = 1 / (k - 1) * torch.sum(k * d - trace)
+    return trace.item()
+
+
+def hutch_plus_plus(A: SymmetricLinOp, k: int, sketch: str) -> float:
+    """Estimate the trace of a symmetric linear operator using the Hutch++ estimator.
+
+    Hutch++ combines sketching and the Hutchinson approach to improve trace estimation.
+
```
Comment (pratikrathore8) [inline rlaopt/spectral_estimators/trace.py:36]: We should add a citation to the paper by Raphael Mayer in the docstring

File: rlaopt/spectral_estimators/trace.py [hunk 4]
```diff
@@ -9,16 +9,53 @@
...
+    Args:
+        A (SymmetricLinOp): The symmetric linear operator.
+        k (int): Number of probe vectors (sketch size).
+        sketch (str): Identifier for the random sketching method.
+
+    Returns:
+        float: Estimated trace of A.
+    """
+    s_dim = k // 2
```
Comment (pratikrathore8) [inline rlaopt/spectral_estimators/trace.py:45]: I think it would be a good idea to allow the user to specify the fraction of vectors to use for trace_sketch vs. trace_hutch (similar to what you showed Zach and I last week). What do you think?

File: rlaopt/spectral_estimators/trace.py [hunk 5]
```diff
@@ -9,16 +9,53 @@
...
+    g_dim = k - s_dim
+    S = get_sketch(sketch, "right", k, A.shape[0], torch.float32, A.device)
```
Comment (pratikrathore8) [inline rlaopt/spectral_estimators/trace.py:47]: We should use `A.dtype` rather than `torch.float32` to avoid precision issues

File: rlaopt/spectral_estimators/trace.py [hunk 6]
```diff
@@ -9,16 +9,53 @@
...
+    AS = S._apply_right(A)
+    Q, _ = torch.linalg.qr(AS, mode="reduced")  # Economic QR
+
+    G = get_sketch(sketch, "right", k, A.shape[0], torch.float32, A.device)
```
Comment (pratikrathore8) [inline rlaopt/spectral_estimators/trace.py:51]: See comment above about `A.dtype`

File: rlaopt/spectral_estimators/trace.py (hunk 4 above)
Comment (oliversssf2) [reply rlaopt/spectral_estimators/trace.py:45]: image

I think the fraction corresponds to the m/3 in the paper, where m is the sketch dimension. I agree that it is kind of weird for the authors to only use 2/3 of the given sketch dimension but this is how they implemented it. Also, the matrix S and G always have the same dimensions in the original implementation but I am happy to make it adjustable in our implementation. What do you think?

Comment (pratikrathore8) [reply rlaopt/spectral_estimators/trace.py:45]: To clarify: if `qS` is the number of queries used for `S` and the fraction of queries used for `G` is `qG`, then the number of total queries is 2 * `qS` + `qG`. The reasoning for this expressions is that the matrix `Q` has the same size as `S`, and therefore, uses the same number of queries as `S`.

My idea is that the sketch size fraction should control the number of vectors used in `S` -- by default the fraction would be set to 1/3. The fraction used for `G` would then be 1 - 2 * sketch size fraction. The size of `Q` would be automatically determined by sketch size fraction, since it is based on the QR decomposition of `A @ S`.

Comment (oliversssf2) [reply rlaopt/spectral_estimators/trace.py:45]: Got it! In this case I would need to force the fraction to be < 1/2 right, or the dimension for Q R would be 0. Would it be okay if the fraction = 1/2, we set the dimension of G to dx1?

Comment (pratikrathore8) [reply rlaopt/spectral_estimators/trace.py:45]: Good point -- thanks for thinking about this!

For now, I think we can raise a `ValueError` if the fraction is greater than or equal to 1/2, and tell the user that the fraction must be smaller than 1/2.

We might also want to check whether `s_dim` or `g_dim` is ever 0. This could occur if the user is using a fraction that is too small, or if the input dimension of the matrix is tiny. When `s_dim` or `g_dim` is 0, I think the right thing to do would be to not compute the trace corresponding to the appropriate component (`trace_sketch` and/or `trace_hutch`) and just set that part to 0.

I'd recommend you update the tests to check all of these edge cases.

Comment (oliversssf2) [reply rlaopt/spectral_estimators/trace.py:45]:
```   python
trace_sketch = 0.0
    trace_hutch = 0.0

    if s_dim > 0:
        S = get_sketch(sketch, "right", s_dim, A.shape[0], A.dtype, A.device)
        AS = S._apply_right(A)
        Q, _ = torch.linalg.qr(AS, mode="reduced")  # Economic QR
        AQ = A @ Q
        trace_sketch = torch.trace(Q.T @ AQ)
    else:
        Q = None  # For safe code path

    if g_dim > 0:
        G = get_sketch(sketch, "right", g_dim, A.shape[0], A.dtype, A.device)
        if s_dim > 0:
            QT_G = G._apply_right(Q.T)
            Q_QT_G = Q @ QT_G
            G_proj = G.Omega_mat - Q_QT_G
        else:
            G_proj = G.Omega_mat

        AG = A @ G_proj
        trace_hutch = torch.trace(G_proj.T @ AG) / g_dim
    # else: Keep trace_hutch = 0.0
    
    return (trace_sketch + trace_hutch).item()
    ```
    
    IS this what you mean?

Comment (oliversssf2) [reply rlaopt/spectral_estimators/trace.py:45]:
```
    if sketch_fraction >= 1/2:
        raise ValueError("sketch_fraction must be smaller than 1/2!!")

    s_dim = k // 2
    g_dim = k - s_dim

    trace_sketch = torch.tensor(0.0, dtype=A.dtype, device=A.device)
    trace_hutch = torch.tensor(0.0, dtype=A.dtype, device=A.device)

    if s_dim > 0:
        S = get_sketch(sketch, "right", s_dim, A.shape[0], A.dtype, A.device)
        AS = S._apply_right(A)
        Q, _ = torch.linalg.qr(AS, mode="reduced")  # Economic QR
        AQ = A @ Q
        trace_sketch = torch.trace(Q.T @ AQ)
    else:
        Q = None  # For safe code path

    if g_dim > 0:
        G = get_sketch(sketch, "right", g_dim, A.shape[0], A.dtype, A.device)
        if s_dim > 0:
            QT_G = G._apply_right(Q.T)
            Q_QT_G = Q @ QT_G
            G_proj = G.Omega_mat - Q_QT_G
        else:
            G_proj = G.Omega_mat

        AG = A @ G_proj
        trace_hutch = torch.trace(G_proj.T @ AG) / g_dim
    # else: Keep trace_hutch = 0.0

    return (trace_sketch + trace_hutch).item()
```

This is what you mean by setting the component to zero when s_dim or g_dim is zero?

Comment (pratikrathore8) [reply rlaopt/spectral_estimators/trace.py:45]: Looking at your code, I think you have the right idea. My only recommendation would be to check `if Q is not None` as opposed to `if s_dim > 0` in the `if g_dim > 0` code block.

I think it's fine to use `math.floor` for now. The other option (which might be simpler) is to throw an error is `s_dim` is 0, and tell the user to increase `sketch_fraction`. I'll leave it to you to decide which approach your prefer
